src/control/cmd/daos_agent/procmon.go

```diff
@@ -0,0 +1,212 @@
+	if !ok {
+		return 0, fmt.Errorf("Underlying type of FileInfo.Sys() is not Stat_t")
+	}
+	return stat.Ino, nil
+}
+
+// monitorProcess is the top level interface for monitoring individual processes.
+// the interface is used by procMon to kick off monitoring individual processes
+// under their own child context to allow for terminating individual monitoring routines.
+func (p *procInfo) monitorProcess() {
+
+	Ino, err := getProcPidInode(p.pid)
+	if err != nil {
+		p.response <- NewResponse(p.pid, err)
+		return
+	}
+
+	p.log.Debugf("Monitoring pid:%d\n", p.pid)
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-time.After(3 * time.Second):
+			newIno, newErr := getProcPidInode(p.pid)
+			if newErr != nil {
+				if os.IsNotExist(newErr) {
+					p.response <- NewResponse(p.pid, fmt.Errorf("Pid %d terminated unexpectedly", p.pid))
+				} else {
+					p.response <- NewResponse(p.pid, newErr)
+				}
+				return
+			}
+			if newIno != Ino {
+				p.response <- NewResponse(p.pid, fmt.Errorf("Pid %d terminated but another processes took its place %d:%d", p.pid, Ino, newIno))
+				return
+			}
+		}
+	}
+}
+
+// procMon is the top level process monitoring struct which accepts requests to
+// monitor and disconnect processes. Once created it is started by passing a
+// context into the startMonitoring call.
+type procMon struct {
+	log     logging.Logger
+	ctx     context.Context
+	procs   map[int32]*procInfo
+	request chan *procMonRequest
+}
+
+// NewProcMon creates a new process monitor struct setting initializing the
+// internal process map and the request channel.
+func NewProcMon(logger logging.Logger) *procMon {
+	return &procMon{
+		log:     logger,
+		procs:   make(map[int32]*procInfo),
+		request: make(chan *procMonRequest),
+	}
+}
+
+func (p *procMon) handleProcessAttach(ctx context.Context, request *procMonRequest, response chan *procMonResponse) {
+	info, ok := p.procs[request.pid]
+
+	if ok == true {
+		p.log.Errorf("Attempted to monitor process %d more than once", request.pid)
+		return
+	}
+
+	info = NewProcInfo(ctx, p.log, request.pid, response)
+
+	p.procs[request.pid] = info
+	go info.monitorProcess()
+}
+
+func (p *procMon) handleProcessDisconnect(request *procMonRequest) {
+	info, ok := p.procs[request.pid]
+
+	if ok != true {
+		p.log.Errorf("Attempted to disconnect process %d but process is missing", request.pid)
+	}
+
+	info.cancelCtx()
+	delete(p.procs, request.pid)
+	p.log.Debugf("Process %d has disconnected", info.pid)
+}
+
+func (p *procMon) handleRequests(ctx context.Context) {
+
+	response := make(chan *procMonResponse)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case request := <-p.request:
+			if request.action == drpc.MethodGetAttachInfo.ID() {
+				p.handleProcessAttach(ctx, request, response)
+
+			} else if request.action == drpc.MethodDisconnect.ID() {
+				p.log.Debugf("Received request in handleRequests to disconnect pid:%d\n", request.pid)
+				p.handleProcessDisconnect(request)
+			}
```

Comment:
Would be cleaner and more maintainable to make this a switch, with a default case logging an error for an unknown action. Also, probably better to log the disconnect in `handleProcessDisconnect()`.

Reply:
Changed to a switch and moved the debug message into the handler.